File: dart/bracket-push/lib/bracket_push.dart
/*
 * copyright 2019 viraltaco_
 * SPDX-License-Identifier: MIT
 * <http://www.opensource.org/licenses/MIT>
 */

class BracketPush {

  /*
   * Inspired by D3Portillo's solution.
   * <https://exercism.io/tracks/dart/exercises/bracket-push/solutions/7fde3b81887b423eb5ba82e21b74218b>
   */
  bool isPaired(String str)
  => str.replaceAll(new RegExp(r'[^\{\}\[\]\(\)]'), '')
        .replaceAll(new RegExp(r'\{\}|\(\)|\[\]'), '')
        .isEmpty
  ;
}

File: cpp/matching-brackets/bracket_push.cpp
/*
 * Copyright (c) 2019 ViralTaco_
 * SPDX-License-Identifier: MIT 
 * <http://www.opensource.org/licenses/MIT>
 */

#include "bracket_push.h"

bool bracket_push::check(const std::string& str)
{
  std::string symbols;
  
  for (const auto c: str) {
    switch (c) {
      case '(': case '{': case '[':
        symbols += c;
        break;
        
      case ')':
        // Calling `back()` on an empty container is undefined.
        if (!symbols.empty() && symbols.back() == '(') {
          symbols.pop_back();
          break;
        } 
        return false;
      case ']':
        if (!symbols.empty() && symbols.back() == '[') {
          symbols.pop_back();
          break;
        }
        return false;
      case '}':
        if (!symbols.empty() && symbols.back() == '{') {
          symbols.pop_back();
          break;
        }
        return false;
        
      default:
        break;
    }
  }
  
  return symbols.empty(); 
}